### PR TITLE
#27 add permissions overview

### LINKED ISSUE #27: Client: Add permissions overview
* Only members with certain permissions should be able to access page
* Display all members having a certain permission
* Handle creation an deletion of permissions

### PR DESCRIPTION
**Pull request by BorchBe99**
closes #27

### CODE REVIEW

File: client/src/ts/members/PermissionsOverview.tsx [hunk 1]
```diff
@@ -0,0 +1,315 @@
...
+      })
+      .then((res) => {
+        if (res.status === 200) {
+          if (mounted) {
+            setPermissionsOverview(res.data);
+          }
+        }
+      })
+      .catch((error) => {
+        console.log(error);
```
Comment (lisa-bai) [inline client/src/ts/members/PermissionsOverview.tsx:132]: Calls to console.log are not allowed

File: client/src/ts/members/PermissionsOverview.tsx [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+      })
+      .then((res) => {
+        if (res.status === 200) {
+          if (mounted) {
+            setMemberPermissions(res.data);
+          }
+        }
+      })
+      .catch((error) => {
+        console.log(error);
```
Comment (lisa-bai) [inline client/src/ts/members/PermissionsOverview.tsx:155]: Calls to console.log are not allowed

File: client/src/ts/members/PermissionsOverview.tsx [hunk 3]
```diff
@@ -0,0 +1,315 @@
...
+        "/users/permissions",
+        { memberID, permissionID },
+        {
+          headers: { Authorization: `Bearer ${localStorage.getItem("token")}` },
+        }
+      )
+      .then((res) => {
+        if (res.status === 200) {
+          if (mounted) {
+            console.log(res.data);
```
Comment (lisa-bai) [inline client/src/ts/members/PermissionsOverview.tsx:177]: Calls to console.log are not allowed

File: client/src/ts/members/PermissionsOverview.tsx [hunk 4]
```diff
@@ -0,0 +1,315 @@
...
+          }
+        }
+      })
+      .catch((error) => {
+        console.log(error);
```
Comment (lisa-bai) [inline client/src/ts/members/PermissionsOverview.tsx:182]: Calls to console.log are not allowed

File: client/src/ts/members/PermissionsOverview.tsx [hunk 5]
```diff
@@ -0,0 +1,315 @@
...
+    api
+      .delete("/users/permissions", {
+        data: { memberID, permissionID },
+
+        headers: { Authorization: `Bearer ${localStorage.getItem("token")}` },
+      })
+      .then((res) => {
+        if (res.status === 200) {
+          if (mounted) {
+            console.log(res.data);
```
Comment (lisa-bai) [inline client/src/ts/members/PermissionsOverview.tsx:202]: Calls to console.log are not allowed

File: client/src/ts/members/PermissionsOverview.tsx [hunk 6]
```diff
@@ -0,0 +1,315 @@
...
+          }
+        }
+      })
+      .catch((error) => {
+        console.log(error);
```
Comment (lisa-bai) [inline client/src/ts/members/PermissionsOverview.tsx:207]: Calls to console.log are not allowed

File: client/src/ts/members/PermissionsOverview.tsx [hunk 7]
```diff
@@ -0,0 +1,315 @@
...
+  const classes = useStyles();
+  const [memberPermissions, setMemberPermissions] = useState<MemberPermissions[]>([]);
+  const [permissionsOverview, setPermissionsOverview] = useState<PermissionsOverview[]>([]);
+  const memberNames = memberPermissions
+    .filter((e) => e.memberID > 0)
+    .map((p) => ({ name: p.name, memberID: p.memberID }))
+    .filter(
+      (value, index, self) => index === self.findIndex((t) => t.memberID === value.memberID && t.name === value.name)
+    );
+  const { auth, dispatchAuth } = useContext(AuthContext);
```
Comment (lisa-bai) [inline client/src/ts/members/PermissionsOverview.tsx:111]: unused constant dispatchAuth

File: client/src/ts/members/PermissionsOverview.tsx (hunk 2 above)
Comment (BorchBe99) [reply client/src/ts/members/PermissionsOverview.tsx]: Fixed

File: client/src/ts/members/PermissionsOverview.tsx (hunk 3 above)
Comment (BorchBe99) [reply client/src/ts/members/PermissionsOverview.tsx]: Fixed

File: client/src/ts/members/PermissionsOverview.tsx (hunk 4 above)
Comment (BorchBe99) [reply client/src/ts/members/PermissionsOverview.tsx]: Fixed

File: client/src/ts/members/PermissionsOverview.tsx (hunk 5 above)
Comment (BorchBe99) [reply client/src/ts/members/PermissionsOverview.tsx]: Fixed

File: client/src/ts/members/PermissionsOverview.tsx (hunk 6 above)
Comment (BorchBe99) [reply client/src/ts/members/PermissionsOverview.tsx]: Fixed

File: client/src/ts/members/PermissionsOverview.tsx (hunk 7 above)
Comment (BorchBe99) [reply client/src/ts/members/PermissionsOverview.tsx]: Fixed

**Review by jbt-it — CHANGES_REQUESTED** [review]
-Please add a it to the navigation under "Mitglieder"
-"createPermission" should also return the changed state?

**Comment by BorchBe99** [comment]
Overview of how Permissions work at current sate:
- The permissions overview  page can only be seen by directors and members with at least one granted permission.
- Members can be applied a director role, but not on this page.
- Only directors can grant and revoke permissions only to other members, if canDeligate is set to 1.
- Members can't grant or revoke permissions.
- Directors can't be granted or revoked any permission.

A test database with a txt file has been uploded to NextCloud

**Review by LennartRukower — APPROVED** [review]
Very good! This can finally be merged :)

**Pull request merged**